snap/quota/quota.go

```diff
@@ -0,0 +1,327 @@
+		}
+	}
+
+	return nil
+}
+
+// NewSubGroup creates a new sub group under the current group.
+func (grp *Group) NewSubGroup(name string, memLimit quantity.Size) (*Group, error) {
+	subGrp := &Group{
+		Name:        name,
+		MemoryLimit: memLimit,
+		ParentGroup: grp.Name,
+		parentGroup: grp,
+	}
+
+	// check early that the sub group name is not the same as that of the
+	// parent, this is fine in systemd world, but in snapd we want unique quota
+	// groups
+	if name == grp.Name {
+		return nil, fmt.Errorf("cannot use same name %q for sub group as parent group", name)
+	}
+
+	if err := subGrp.validate(); err != nil {
+		return nil, err
+	}
+
+	// save the details of this new sub-group in the parent group
+	grp.subGroups = append(grp.subGroups, subGrp)
+	grp.SubGroups = append(grp.SubGroups, name)
+
+	return subGrp, nil
+}
+
+// ResolveCrossReferences takes a set of deserialized groups and sets all
+// cross references amongst them using the unexported fields which are not
+// serialized.
+func ResolveCrossReferences(grps map[string]*Group) error {
+	// iterate over all groups, looking for sub-groups which need to be threaded
+	// together with their respective parent groups from the set
+	for name, grp := range grps {
+		if name != grp.Name {
+			return fmt.Errorf("group has name %q, but is referenced as %q", grp.Name, name)
+		}
+
+		// validate the group, assuming it is unresolved
+		if err := grp.validate(); err != nil {
+			return fmt.Errorf("group %q is invalid: %v", name, err)
+		}
+
+		// first thread the parent link
+		if grp.ParentGroup != "" {
+			parent, ok := grps[grp.ParentGroup]
+			if !ok {
+				return fmt.Errorf("missing group %q referenced as the parent of group %q", grp.ParentGroup, grp.Name)
+			}
+			grp.parentGroup = parent
+
+			// make sure that the parent group references this group
+			found := false
+			for _, parentChildName := range parent.SubGroups {
+				if parentChildName == grp.Name {
+					found = true
+				}
+			}
+			if !found {
+				return fmt.Errorf("group %q does not reference necessary child group %q", parent.Name, grp.Name)
+			}
+		}
+
+		// now thread any child links from this group to any children
+		if len(grp.SubGroups) != 0 {
+			grp.subGroups = make([]*Group, len(grp.SubGroups))
+			for i, subName := range grp.SubGroups {
+				sub, ok := grps[subName]
+				if !ok {
+					return fmt.Errorf("missing group %q referenced as the sub-group of group %q", subName, grp.Name)
+				}
+
+				// check that this sub-group references this group as it's
+				// parent
+				if sub.ParentGroup != grp.Name {
+					return fmt.Errorf("group %q does not reference necessary parent group %q", sub.Name, grp.Name)
+				}
+
+				grp.subGroups[i] = sub
+			}
+		}
+	}
+
+	return nil
+}
+
+// tree recursively returns all of the sub-groups of the group and the group
+// itself.
+func (grp *Group) tree() []*Group {
+	// TODO: should we be paranoid about circular references and
+	// recursion here?
```

Comment:
I think should do something about it (may come later).

Reply:
I added something in https://github.com/snapcore/snapd/pull/10198